### PR TITLE
introduce dns events

### PR DESCRIPTION
**Pull request by itamarmaouda101**
Hi 
This PR solves #1384.

Its adds 2 new events that are part of the net events.
Also, it adds the set of the net events to the `events_definitions `map and changes the function that handles the net events to be able to handle more than one event (until today we had only one net event which is `NetPacket`).

How it works:
In the `tc_probe` function at the bpf code, I check if the packet is has a src/dst port of DNS (53) and if it's UDP.
Then it sends it with event_id of `DnsRequest` or `DnsResponse`.
in the `net_event.go` I added a function that will handle the processing of all the network events.
It basically has a switch-case with the event_id of the packet and the net events types, in case it found the event, it calls its `event_handler_function`.

Here's an example for output for trying loacte the ip of www.ubuntu.com:
`itamar@ubuntu:~/golang_env$ sudo ./dist/tracee-ebpf -t  event=dns_request,dns_response -t net=ens33 -o json
`
libbpf: Kernel error message: Exclusivity flag on, cannot modify

"timestamp":1649334493999995757,"processorId":0,"processId":759,"threadId":759,"parentProcessId":759,"hostProcessId":759,"hostThreadId":759,"hostParentProcessId":1,"userId":101,"mountNamespace":4026532596,"pidNamespace":4026531836,"processName":"systemd-resolve\u0000","hostName":"","containerId":"","eventId":"4001","eventName":"dns_request","argsNum":2,"returnValue":0,"stackAddresses":null,"args":[{"name":"metadata","type":"trace.PktMeta","value":{"src_ip":"192.168.16.134","dst_ip":"192.168.16.2","src_port":35807,"dst_port":53,"protocol":17}},{"name":"dns_questions","type":"[]bufferdecoder.DnsQueryData","value":[{"query":"www.ubuntu.com","queryType":"A","queryClass":"IN"}]}]}

{"timestamp":1649334494085087526,"processorId":0,"processId":759,"threadId":759,"parentProcessId":759,"hostProcessId":759,"hostThreadId":759,"hostParentProcessId":1,"userId":101,"mountNamespace":4026532596,"pidNamespace":4026531836,"processName":"systemd-resolve\u0000","hostName":"","containerId":"","eventId":"4002","eventName":"dns_response","argsNum":2,"returnValue":0,"stackAddresses":null,"args":[{"name":"metadata","type":"trace.PktMeta","value":{"src_ip":"192.168.16.2","dst_ip":"192.168.16.134","src_port":53,"dst_port":35807,"protocol":17}},{"name":"dns_response","type":"[]bufferdecoder.DnsResponseData","value":[{"query_data":{"query":"www.ubuntu.com","queryType":"A","queryClass":"IN"},"dns_answer":[{"answer_type":"A","ttl":5,"answer":"185.125.190.20"},{"answer_type":"A","ttl":5,"answer":"185.125.190.21"},{"answer_type":"A","ttl":5,"answer":"185.125.190.29"}]}]}]}

{"timestamp":1649334494088166734,"processorId":0,"processId":759,"threadId":759,"parentProcessId":759,"hostProcessId":759,"hostThreadId":759,"hostParentProcessId":1,"userId":101,"mountNamespace":4026532596,"pidNamespace":4026531836,"processName":"systemd-resolve\u0000","hostName":"","containerId":"","eventId":"4001","eventName":"dns_request","argsNum":2,"returnValue":0,"stackAddresses":null,"args":[{"name":"metadata","type":"trace.PktMeta","value":{"src_ip":"192.168.16.134","dst_ip":"192.168.16.2","src_port":51095,"dst_port":53,"protocol":17}},{"name":"dns_questions","type":"[]bufferdecoder.DnsQueryData","value":[{"query":"www.ubuntu.com","queryType":"AAAA","queryClass":"IN"}]}]}

{"timestamp":1649334494169898401,"processorId":0,"processId":759,"threadId":759,"parentProcessId":759,"hostProcessId":759,"hostThreadId":759,"hostParentProcessId":1,"userId":101,"mountNamespace":4026532596,"pidNamespace":4026531836,"processName":"systemd-resolve\u0000","hostName":"","containerId":"","eventId":"4002","eventName":"dns_response","argsNum":2,"returnValue":0,"stackAddresses":null,"args":[{"name":"metadata","type":"trace.PktMeta","value":{"src_ip":"192.168.16.2","dst_ip":"192.168.16.134","src_port":53,"dst_port":51095,"protocol":17}},{"name":"dns_response","type":"[]bufferdecoder.DnsResponseData","value":[{"query_data":{"query":"www.ubuntu.com","queryType":"AAAA","queryClass":"IN"},"dns_answer":[{"answer_type":"AAAA","ttl":5,"answer":"2620:2d:4000:1::26"},{"answer_type":"AAAA","ttl":5,"answer":"2620:2d:4000:1::28"},{"answer_type":"AAAA","ttl":5,"answer":"2620:2d:4000:1::27"}]}]}]}

### CODE REVIEW

File: pkg/ebpf/c/tracee.bpf.c [hunk 1]
```diff
@@ -4836,6 +4838,21 @@ static __always_inline bool skb_revalidate_data(struct __sk_buff *skb, uint8_t *
...
+void set_net_event_id(net_packet_t *pkt){
+    if (pkt->protocol == IPPROTO_UDP && pkt->dst_port == 53){
+        pkt->event_id = DNS_REQUEST;
+    }
+    if (pkt->protocol == IPPROTO_UDP && pkt->src_port == 53){
+        pkt->event_id = DNS_RESPONSE;
+    }
+}
+
+bool should_send_payload(net_packet_t *pkt){
```
Comment (AlonZivony) [inline pkg/ebpf/c/tracee.bpf.c:4850]: I think that the word we use to describe sending data from kernel to user is "submit" so I think you should call this like this.

File: pkg/ebpf/tracee.go [hunk 2]
```diff
@@ -1293,6 +1293,14 @@ func findInList(element string, list *[]string) (int, error) {
 	}
 	return 0, fmt.Errorf("element: %s dosent found\n", element)
 }
+func (t *Tracee) netEventChoosen() bool {
```
Comment (AlonZivony) [inline pkg/ebpf/tracee.go:1296]: Change it to `shouldTraceNetEvents`

File: pkg/bufferdecoder/net_decoder.go [hunk 3]
```diff
@@ -118,3 +129,74 @@ func (decoder *EbpfDecoder) DecodeNetDebugEvent(netDebugEvent *NetDebugEvent) er
...
+	return request
+}
+
+func (decoder *EbpfDecoder) DecodeDnsQueryArray(questions *[]DnsQueryData) error {
+	offset := decoder.cursor
+	packet := gopacket.NewPacket(decoder.buffer[offset:], layers.LayerTypeEthernet, gopacket.Default)
+	if packet == nil {
+		return fmt.Errorf("couldnt parse the packet")
+	}
+	dnsLayer := packet.Layer(layers.LayerTypeDNS).(*layers.DNS)
```
Comment (AlonZivony) [inline pkg/bufferdecoder/net_decoder.go:154]: I think that a cleaner way to do this is to use the following format:
```suggestion
	dnsLayer, ok := packet.Layer(layers.LayerTypeDNS).(*layers.DNS)
	if !ok {
```

File: pkg/bufferdecoder/net_decoder.go [hunk 4]
```diff
@@ -32,6 +34,15 @@ func (decoder *EbpfDecoder) DecodeNetEventMetadata(eventMetaData *NetEventMetada
 	return nil
 }
 
+func (decoder *EbpfDecoder) DecodePacketBytes(packetBytes []byte, packetstartOffset uint32, packetLength uint32) error {
+	if int(packetstartOffset+packetLength) > len(decoder.buffer) {
+		print(packetstartOffset+packetLength, " len is ", decoder.buffer)
+		return fmt.Errorf("offset to the payload to big")
+	}
+	_ = copy(packetBytes[:], decoder.buffer[packetstartOffset:packetstartOffset+packetLength])
```
Comment (AlonZivony) [inline pkg/bufferdecoder/net_decoder.go:42]:
```suggestion
	_ = copy(packetBytes[:], decoder.buffer[ packetstartOffset: (packetstartOffset + packetLength)])
```

**Review by AlonZivony — CHANGES_REQUESTED** [review]
Good work.
Small changes for the code to match project standards.

File: pkg/ebpf/net_events.go [hunk 5]
```diff
@@ -233,23 +232,27 @@ func (t *Tracee) processNetEvents(ctx gocontext.Context) {
...
 				if err != nil {
 					t.handleError(err)
 					continue
 				}
+				PacketStartOff := netEventMetadata.GetSizeBytes() + netCaptureData.GetSizeBytes()
 
 				ifaceName := t.netInfo.ifaces[int(netCaptureData.ConfigIfaceIndex)].Name
 				ifaceIdx, err := t.getTracedIfaceIdx(ifaceName)
 				if err == nil && ifaceIdx >= 0 {
+					PacketStartOff += 40 // the offset to the packet payload the size of the NetEventMetadata + NetPacketEvent
```
[the remark this answers is not in the document]
Comment (itamarmaouda101) [reply pkg/ebpf/net_events.go:247]: yep , put the on of the comment in incorrect place, just fixed that (:

File: pkg/ebpf/net_events.go [hunk 6]
```diff
@@ -233,23 +233,27 @@ func (t *Tracee) processNetEvents(ctx gocontext.Context) {
...
 
-			if netEventMetadata.NetEventId == NetPacket {
+			if isNetEvent(netEventMetadata.NetEventId) {
 				var netCaptureData bufferdecoder.NetCaptureData
 				err = netDecoder.DecodeNetCaptureData(&netCaptureData)
 				if err != nil {
 					t.handleError(err)
 					continue
 				}
+				PacketStartOff := netEventMetadata.GetSizeBytes() + netCaptureData.GetSizeBytes() // packet payload offset is NetEventMetadata + NetCaptureData
```
Comment (AlonZivony) [inline pkg/ebpf/net_events.go:243]: The comment here is the same as the code. It has no added value.
Please write in the comment why that packet start after both of this structs (maybe put it above the line).

File: pkg/ebpf/net_events.go [hunk 7]
```diff
@@ -233,23 +233,27 @@ func (t *Tracee) processNetEvents(ctx gocontext.Context) {
...
 
 				ifaceName := t.netInfo.ifaces[int(netCaptureData.ConfigIfaceIndex)].Name
 				ifaceIdx, err := t.getTracedIfaceIdx(ifaceName)
 				if err == nil && ifaceIdx >= 0 {
+					PacketStartOff += netPacketSize // if the interface's packet should be trace we need to add another the size of NetPacketEvent
```
Comment (AlonZivony) [inline pkg/ebpf/net_events.go:248]:
```suggestion
					 // If the interface's packet should be traced than NetPacketEvent fields are added to submitted buffer.
					 // Start offset should be updated accordingly 
					PacketStartOff += netPacketSize
```

**Review by AlonZivony — CHANGES_REQUESTED** [review]
Some doc changes required

**Comment by rafaeldtinoco** [comment]
Closing this on behalf of https://github.com/aquasecurity/tracee/pull/1698 being taken care by @roikol now.

**Pull request closed without merging**